File: bookstore-micronaut-example/bookstore-service/src/main/java/example/bookstore/micronaut/api/controller/BookController.java
package example.bookstore.micronaut.api.controller;

import example.bookstore.micronaut.api.model.Book;
import example.bookstore.micronaut.api.model.BookRequest;
import example.bookstore.micronaut.api.model.search.BookSearchCriteria;
import example.bookstore.micronaut.api.model.search.PagedSearchResult;
import example.bookstore.micronaut.core.service.BookService;
import io.micronaut.http.MediaType;
import io.micronaut.http.annotation.*;
import io.swagger.v3.oas.annotations.tags.Tag;
import jakarta.inject.Inject;

@Controller(value = "/books", produces = MediaType.APPLICATION_JSON)
@Tag(name = "book")
public class BookController {

    @Inject
    private BookService bookService;

    @Get
    public PagedSearchResult<Book> getBooks(@RequestBean BookSearchCriteria criteria) {
        return bookService.getBooks(criteria);
    }

    @Get("/{bookId}")
    public Book getBook(Long bookId) {
        return bookService.getBook(bookId);
    }

    @Post
    public Book createBook(@Body BookRequest request) {
        return bookService.createBook(request);
    }

    @Put("/{bookId}")
    public Book updateBook(Long bookId, @Body BookRequest request) {
        return bookService.updateBook(bookId, request);
    }
}

File: bookstore-service-micronaut-kotlin/src/main/kotlin/com/example/micronaut/bookstore/controller/BookController.kt
package com.example.micronaut.bookstore.controller

import com.example.micronaut.bookstore.controller.model.Book
import com.example.micronaut.bookstore.controller.model.CreateBookRequest
import com.example.micronaut.bookstore.service.BookService
import com.example.micronaut.bookstore.service.model.BookSearchCriteria
import io.micronaut.data.model.Page
import io.micronaut.http.MediaType
import io.micronaut.http.annotation.*

@Controller(value = "/books", produces = [MediaType.APPLICATION_JSON])
class BookController(private val bookService: BookService) {

    @Get
    fun getBooks(@RequestBean bookSearchCriteria: BookSearchCriteria): Page<Book> {
        return bookService.getBooks(bookSearchCriteria)
    }

    @Post
    fun createBook(@Body createBookRequest: CreateBookRequest): Book {
        return bookService.createBook(createBookRequest)
    }
}